lib/react.js

```diff
@@ -269,6 +271,58 @@ declare module react {
     toArray<T>(children: ChildrenArray<T>): Array<$NonMaybeType<T>>;
   };
 
+  declare type BasicStateAction<S> = (S => S) | S;
+  declare type Dispatch<A> = A => void;
+  declare type MaybeCleanUpFn = ?(() => mixed);
+
+  declare export function useContext<T>(
+    context: React$Context<T>,
+    observedBits: void | number | boolean,
+  ): T;
+
+  declare export function useState<S>(
+    initialState: (() => S) | S,
+  ): [S, Dispatch<BasicStateAction<S>>];
+
+  declare export function useReducer<S, A>(
+    reducer: (S, A) => S,
+    initialState: S,
+    initialAction: ?A,
+  ): [S, Dispatch<A>];
+
+  declare export function useRef<T>(initialValue: ?T): {current: T | null};
+
+  declare export function useEffect(
+    create: () => MaybeCleanUpFn,
+    inputs: ?$ReadOnlyArray<mixed>,
+  ): void;
+
+  declare export function useMutationEffect(
+    create: () => MaybeCleanUpFn,
+    inputs: ?$ReadOnlyArray<mixed>,
+  ): void;
+
+  declare export function useLayoutEffect(
+    create: () => MaybeCleanUpFn,
+    inputs: ?$ReadOnlyArray<mixed>,
+  ): void;
+
+  declare export function useCallback<T>(
+    callback: () => T | void,
```

Comment:
@bvaughn Hi.  Is this supposed that  callback function is always without args?
`const onSubmit = useCallback((values) => {}); return <Form onSubmit={onSubmit} />` - how should we type this case, where values passed as args to the callback from inside child component?

Reply:
Values are passed from parent scope. Nothing is passed to callback.
```
const onSubmit = useCallback(() => {
  fetch('/update', id)
  // this is only invalidation param
}, [id])
```

Reply:
As I said, i need to pass callback which accept args from child component. It doens't make sense that its impossible.

something like that.

https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBLAtgBzgJwBdkwBDAZzACUBTUgY2KnzizAHJ87H2BuTXAWIBvMAFdyNAKJQoNRmAC+YZqw5cGhPunpwAduWIAFUlz3EAvGAAUASjAWAfGGGowYXQeL6AymIBGWBiW1gBupDBiNOT2Ti5u7mDAwHKE9AAWCYoJCVyEYvh6YAA8AGIEbL4BQYQWwlWBwcrAjqiK-KiehmDl+GxW1vV6fo2EirHOru7FUBVgDTV1C02tiSW9WKUYNDAAJmB6pFg0FgBEs3CnSaslKRWtikA

it can be form onSubmit with values..or whatever other callback with args we can imagine.

its absolutely valid from react side, but not from these typings.

Reply:
and there is one more issue with those typings, we can't pass `async` keyword to callbacks in the hooks, like here:
https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBLAtgBzgJwBdkwBDAZzACUBTUgY2KnzizAHJ87H2BuTXAWIBvMAFdyNAKJQoNRmAC+YZqw5cGhPunpwAduWIAxOHDABeMAAoAlBYB8YYajDjJMuYysUAnnvrWduaOwooANGAA2gC6NvyuXIRi+HpgADwAJhgAbmDA9qiKqEA

Reply:
Ah, good catch. The callback function passed to `useCallback`**can** accept parameters. (We have some tests for this in React, e.g. [1](https://github.com/facebook/react/blob/88ada98198b22d2b8cd056cfba5ffb113c6a50d3/packages/react-dom/src/__tests__/ReactDOMServerIntegrationHooks-test.internal.js#L520-L522), [2](https://github.com/facebook/react/blob/88ada98198b22d2b8cd056cfba5ffb113c6a50d3/packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.internal.js#L1494-L1500)).

Reply:
I _believe_ I've fixed the broken `useCallback` definition via e33b4b57e

Reply:
cool, thanks! 👍